asynction/security.py

```diff
@@ -0,0 +1,412 @@
+import base64
+import http.cookies
+from functools import wraps
+from typing import Callable
+from typing import List
+from typing import Mapping
+from typing import Optional
+from typing import Sequence
+from typing import Tuple
+from typing import Union
+
+from flask import Request
+from flask import request as current_flask_request
+
+from asynction.exceptions import SecurityException
+from asynction.types import HTTPAuthenticationScheme
+from asynction.types import SecurityRequirement
+from asynction.types import SecurityScheme
+from asynction.types import SecuritySchemesType
+
+TokenInfoFunc = Callable[[str], Mapping]
+BasicInfoFunc = Callable[[str, str, Optional[Sequence[str]]], Mapping]
+APIKeyInfoFunc = Callable[[str, Optional[Sequence[str]], Optional[str]], Mapping]
+ScopeValidateFunc = Callable[[Sequence[str], Sequence[str]], bool]
+InternalSecurityCheckResponse = Tuple[Optional[Mapping], Optional[str]]
+InternalSecurityCheck = Callable[[Request], InternalSecurityCheckResponse]
+SecurityCheck = Callable[[Request], Mapping]
+
+InternalSecurityRequirement = Tuple[str, Sequence[str]]
+SecurityCheckFactory = Callable[
+    [InternalSecurityRequirement, SecurityScheme], Optional[InternalSecurityCheck]
+]
+
+
+def unpack_security_requirement(
+    requirement: SecurityRequirement,
+) -> InternalSecurityRequirement:
+    return next(iter(requirement.items()))
+
+
+def unpack_security_requirements(
+    requirements: Sequence[SecurityRequirement],
+) -> Sequence[InternalSecurityRequirement]:
+    return list(map(unpack_security_requirement, requirements))
+
+
+def extract_auth_header(request: Request) -> Optional[Tuple[str, str]]:
+    authorization = request.headers.get("Authorization")
+
+    if not authorization:
+        return None
+    try:
+        lhs, rhs = authorization.split(None, 1)
+        return lhs, rhs
+    except ValueError as err:
+        raise SecurityException from err
+
+
+def validate_basic(
+    request: Request, basic_info_func: BasicInfoFunc, required_scopes: Sequence[str]
+) -> Union[Mapping, None]:
```

Comment:
```suggestion
) -> Optional[Mapping]:
```
Could we also make this `Mapping` stricter? Maybe we can use a [TypedDict](https://docs.python.org/3/library/typing.html#typing.TypedDict) and import it from typing-extensions, since it's not available on python3.7.

Reply:
For example:
```python
class SecurityInfo(TypedDict):
    user: str
    scopes: Sequence[str]
```

Reply:
Personally I would prefer not to. With connexion I actually implement my x-tokenInfoFunc to lookup the user by the string and resolve that to the user DB object so that it gets passed into the HTTP handler. Being able to do the same would be better IMO. If there is a way to require that the the scopes variable is as expected but to allow any others to just be forwarded through that would be nice.

Reply:
This is a valid point. However we do expect all security functions to return scopes, so I'd probably go with a generic type that includes scopes along with any other form of data:

```python
T = TypeVar("T")

class SecurityInfo(Generic[T]):
    scopes: Sequence[str]
    data: T

# The dev would then need to something along the lines of:

from asynction import SecurityInfo

@dataclass
class Foo:
    bar: str

def my_token_info_func(token: str) -> SecurityInfo[Foo]:
     # validate stuff here
```
or
```python
class SecurityInfoBase(Protocol): # see typing.protocol
    scopes: Sequence[str]

SecurityInfo = TypeVar("SecurityInfo", bound=SecurityInfoBase)

# The dev would then need to something along the lines of:

from asynction import SecurityInfoBase

@dataclass
class Foo(SecurityInfoBase):
    scopes: Sequence[str]
    bar: str

def my_token_info_func(token: str) -> Foo:
     # validate stuff here
```

Since users of asynction are expected to use this type when building their security validators, we need to expose the type in the top level [__init__](https://github.com/dedoussis/asynction/blob/main/asynction/__init__.py) module.

Reply:
The problem I see with both of those options is that it is diverging from how connexion does things. In my use case I would be handing the same x-tokenInfoFunc to both connexion and asynction. So having different required signatures for each would be annoying and cause code duplication. I see the desire to want this to be more type safe, but in this case it seems like asynction would be imposing it's own philosophies onto the end users code (required strict typing) where the only actual contract that needs to be enforced is that the response is a dict that contains "scope" or "scopes". That is something that is easy to verify at runtime and hard to prove with type checking as we don't have access to the end users code during asynctions build process.

Reply:
@alex-zywicki I do agree with the sentiment of making signatures similar to connexion. To achieve that we can use a `TypedDict` class that allows customisability through inheritance. For example:

```python
from typing import Sequence
from typing_extensions import TypedDict  # need to use typing_extensions for compatibility with python3.7

class SecurityInfo(TypedDict, total=False):
    scopes: Sequence[str]
    scope: str

class FooInfo(SecurityInfo):
    bar: str

def token_info_func(token: str) -> FooInfo:
    return {
        "scope": "read write",
        "bar": "baz"
    }
```
The `SecurityInfo` type could also be used directly as a return annotation, if the user doesn't want to add more data into the dictionary.

With the above approach, we can achieve both signature compatibility and type safety.
I'm keen to have clear typing contracts for the security callables. It's essentially extra testing and documentation for free.

Reply:
@alex-zywicki Well, the fact that the connexion type is even more complex makes the typing argument stronger.

The more fields the `token_info` dict can have, the more important it is to type/document it.
We can just add the `sub` and `uid` str fields next to `scope` and `scopes` in the `SecurityInfo` typed dictionary.

Reply:
And what would the type of the sub and uid fields be? I would think it should be `Any` since there isn't really a function requirement that it be a `str` and my use case would benefit from that.

Reply:
Right, I see. It would be great if we could use generic types for this, but we cannot (TypedDict does not support generics). OK let's add `sub` and `uid` there as `Any`. Unfortunately the `Any` annotation removes any typechecking benefit we could have, but at least we have the fields there for documentation purposes.